### ISSUE TITLE
[css-gcpm-3] string-set with counter(page), when is value resolved?

### ISSUE THREAD

**Issue by faceless2** [w3c/csswg-drafts#4740]
Take the following CSS and HTML:

```css
:root {
  counter-reset: page 0; /* UA rule */
}
@page {
  counter-increment: page 1; /* UA rule */
  @top-center {
    content: "page " string(x, first);
  }
}
body {
  string-set: x counter(page, decimal);
}
```
```html
<body>
  ... multi page content ...
</body>
```
What is displayed at the top of each page? There are two conceivable options:

1. The value is "page 1" on all pages, which is the result you would get it you set the value of "x" to the value of the _page_ counter when the string was first created.

2. The value is "page _n_", where page is the current page number. This is the result you would get if you evaluate the value of the _page_ counter when the string "x" is displayed, not when its set.

Prince chooses option 2, at least for the page counter - this exact approach is used in their "essay.html" example available at https://css4.pub. I am fairly convinced this is wrong, but have realised that this is not explicit in the specification (which is [here](https://drafts.csswg.org/css-gcpm/#setting-named-strings-the-string-set-pro), for reference).

Why do I think this is wrong? Because it is inconsistent with the way other counter values are used. Take an almost identical example:

```css
:root {
  counter-reset: section 0;
}
@page {
  @top-center {
    content: "section " string(x, first);
  }
}
h1 {
  counter-increment: section 1;
  string-set: x counter(section, decimal);
}
```
```html
<body>
  <h1>Header</h1>
  <h1>Header</h1>
  <h1>Header</h1>
</body>
```
The only difference here is that the counter is being incremented on the `<h1>`, not in the page rule. The intention of this is clear, and examples showing this type of approach exist in the spec and in various testcases. All agree that the value at the top of the page is "section 1".

But the only way you can get "section 1" is by evaluating the _section_ counter __when the string is set__. If you evaluate it when the string is displayed, you will get the current value of the counter, which - in this example - would result in "section 3".

The page counter is not magic - it is a normal counter as described in https://drafts.csswg.org/css-page-3/#page-based-counters, and the rules marked "UA rule" in the first example would be in the UA stylesheet, or equivalent structure.

I'm prepared to concede that there is a third way to conceive of this beyond the reach of my feeble mind. Whatever the solution, I think this needs clarification in the specification.

I'll prepare a proper (tentative) testcase and would be interested to hear the results from other paged media renderers.

**Comment by dauwhe** [comment]
I think authors would find it tremendously confusing if the page counter did not display the value of the current page. Why should string-set freeze the value of a counter? 

> But the only way you can get "section 1" is by evaluating the _section_ counter **when the string is set**. If you evaluate it when the string is displayed, you will get the current value of the counter, which - in this example - would result in "section 3".

I don't think this is accurate. The value of the string `x` changes at several places on the page—it is "1" at the start and "3" at the end. You explicitly ask for the first value with `string-set`, so you get "1".

**Comment by dauwhe** [comment]
PrinceXML and WeasyPrint increment the page numbers, PDFReactor does not. Supposedly I installed AH today, but there's no evidence of that on my computer :)

Can you describe some use cases for this behavior?

**Comment by faceless2** [comment]
Without resolving this issue, the behaviour of a string set to _any_ counter incremented in the page margin is quite different to a string set to anything else. Prince only allows one counter in the page margin, `page`, which might be why I'm struggling to convey the issue if that's the one you're testing against. But others might include:

1. Keeping track of the physical number of pages, as Antenna House does with their "page physical" counter.
2. Displaying the number of pages in a page group, or a chapter, or appendix.


Let me try again with a self-contained, complete example. Would be great to get your results from PDFReactor, and AH Formatter if you ever find it :-)

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
 <head>
  <style>
  :root {
     counter-reset: page 0 b 0;
  }
  @page {
     counter-increment: page 1 b 1;
     @top-center {
         content: counter(page) counter(b) string(c) string(d);
     }
  }
  div {
      break-before: page;
  }
  #div1 {
      string-set: c counter(page), d counter(b);
  }
  </style>
 </head>
 <body>
  <div id="div1">Div1</div>
  <div id="div2">Div2</div>
  <div id="div3">Div3</div>
 </body>
</html>
```

What I expect to see here is that `counter(page)` and `counter(b)` increment on each page, but the two strings are set to the value "1" on the first page and remain at that value. So we should have page1=1111, page2=2211, page3=3311 (the final two values could also be "0", that's a separate issue with counter scope I'm about to file against css-page. But the point is, they don't change).

The results I've tested here are:

* Weasyprint: page1=1100, page2=2200, page3=3300
* Prince: page1=1010, page2=2020, page3=3030
* Our unreleased renderer: page1=1111, page2=2211, page3=3311
* PDFReactor: not tested
* Antenna House: not tested


Prince doesn't handle any page-margin counters other than "page", which is why "b" is never incremented. Because the `page` counter is magic, they can easily extend that magic to any string set to the value of `counter(page)` and have it always reflect the current page number.

But once you open the door to counters other than `page` being incremented in page margins it gets much harder. If "page" type counters are handled differently to other counters, you have to keep track of which is which. And if you increment a counter in _both_ contexts - odd, but perfectly valid - you're stuck (I have a pile of tests for this topic, but the good news is they're all against css-page, so you're off the hook for those ones).

I've expended a lot of words on this topic, sorry. My fundamental point is that "the page counter is not magic".  We _have_ to be able to explain the behaviour of `string: x counter(page)` without treating "page" differently to any other counter.

**Comment by dauwhe** [comment]
PDFReactor: page1=1100, page2=2200, page3=3300

No luck with AntennaHouse yet :)

**Comment by faceless2** [comment]
Vivliostyle is unwilling to display a string for me at all sadlly. But the "b" counter is tracking the "page" counter at least.

**Comment by dauwhe** [comment]
paged.js doesn't support `string-set` yet, so I've been unable to test there.

**Comment by dauwhe** [comment]
Counters that increment in `@page` are different from other counters because their values can't be determined until layout has taken place—they don't operate on the DOM but on the fragmentainers of the box tree. 

I do think this is a really interesting question, and would like to get some feedback from some of the core CSSWG folks.

**Comment by faceless2** [comment]
Yes, the value not being determined until after layout makes things quite a lot more complicated (e.g. https://github.com/w3c/csswg-drafts/issues/4760). But once you've got that working, that they're no different to regular counters.

I agree we need another opinion, so I will stop badgering you on this one - for now at least. Perhaps we should ask the css print community group for their opinions?

If you ever find Antenna House Formatter, please do add the results.

I updated https://github.com/w3c/csswg-drafts/issues/4759 to explain why we think the values on the first page should be "1111" not "1100"

**Comment by faceless2** [comment]
@dauwhe not sure the procedure here, but - as we've debated ourselves into a corner on this issue, and we both agree we need further input - I've put it down for the upcoming F2F. That's presuming you're able to make the call, if not  feel free to remove the label.

**Comment by dauwhe** [comment]
I'm not necessarily disagreeing. The spec does say:

> The content values of named strings are assigned at the point when the content box of the element is first created (or would have been created if the element’s display value is none).

I filed an [issue](https://www.princexml.com/forum/topic/4389/using-page-counters-inside-string-function) with Prince, hoping to get their feedback.

**Comment by faceless2** [comment]
I've been trying to figure out why anyone would use `string-set: f, counter(page))` and `content: string(f)` instead of just `content: counter(page)`, as it seems to have no benefit.

The answer is that it lets you do this
```css
@page {
  @top-left {
    content: string(page-number, first-except);
  }
}
section {
    string-set: page-number counter(page);
}
```
Prince's implementation uses this approach with `first-except` to let you _not display_ the page number on the first page of the section - the page where the string was set.

I still think this approach doesn't match what's currently specified. But it explains why it's considered useful, and it's an effect that's difficult to reproduce without this behaviour (you need the `:nth()` page selector). It's also clear from the differing results above that a) the behaviour needs defining one way or another, and b) doing so is going to cause either prince, or everyone else, to change the behaviour.

---

Can we define the behaviour of Prince in a general way so that it's applicable for all counters and doesn't require magic for `page`? Yes I think we can, with something like the following text added to the definition of `string-set`:

> When the value of a string includes a `counter()` or `counters()` function, it is treated as a _reference_ to that counter, not as a copy of the current counter value.

and the following added to the definition of `string()`:

> The second argument to the `string()` function is used to determine not only which _value_ of the string is used, but also - if that chosen value of the string contains a counter reference - which _value_ of the counter is displayed. The _entry value_ for a counter will include any assignments made in the page margin.

With this revised definition, the output from the example above would be page1=1111, page2=222, page3=3333.

I'll test this out next week to see if it works in practice. It would still be really, really useful to see what Antenna House's renderer does with the above example:

@tabatkins I recall you downloaded AH renderer for another issue. Any chance you could try https://github.com/w3c/csswg-drafts/issues/4740#issuecomment-592085773 and update this issue with the results?

**Comment by faceless2** [comment]
The above changes to the definitions seem to match the behaviour of prince's "page" counter, without breaking any of our other tests. It's impossible to match the behaviour of Prince in one corrner, and everyone-but-Prince in the other corner, but  this comes a lot closer. 

Here's another example to test:
```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
 <head>
  <style>
  @page {
     margin: 36px;
     @top-center {
         content: string(b);
     }
  }
  body {
     string-set: b "Value=" counter(b) ", Page=" counter(page);
     counter-reset: b 1;
  }
  div {
      break-before: page;
      counter-increment: b;
  }
  </style>
 </head>
 <body>
  The top of this page is "Value=1, Page=1"
  <div id="div1">
   The top of this page is:
   <ul>
    <li>"Value=1, Page=1" if counters are copied to a string by value.</li>
    <li>"Value=2, Page=2" if counters are copied to a string by reference.</li>
    <li>"Value=1, Page=2" if the page counter is copied by reference and others by value</li>
   </ul>
  </div>
 </body>
</html>
```
Prince currently gives "Value=1, Page=2", because "page" counters are magic in Prince. I expect all other renderers will give "Value=1, Page=1". With this proposed change, the output would be "Value=2, Page=2".

The difference will be the case where a string is set to a counter, then the counter is changed but the string is left unchanged, and finally the string is displayed in the margin. I cannot imagine a situation where this behaviour (an unchanging string in the margins, while the counter it depends upon is changing) would be required.

Another option is to add a parameter to `string-set` which determines whether a counter is copied to the string by value or by reference. But I think that's unnecessarily complicated.
